File: backend/app/routers/leaders.py
import datetime
from typing import Optional

from fastapi import APIRouter

from app.core.serialize import df_to_records, sanitize_floats
from app.db import queries as db
from app.services import leaders as leaders_service

router = APIRouter(prefix="/api/leaders", tags=["leaders"])


@router.get("")
def get_leaders(
    broad_cat: Optional[str] = None,
    sub_cat: Optional[str] = None,
    plan_type: Optional[str] = None,
    option_type: Optional[str] = None,
    search: str = "",
    start: Optional[datetime.date] = None,
    end: Optional[datetime.date] = None,
) -> dict:
    df_all = db.get_advanced_leaders_dataframe(
        broad_cat=broad_cat or "All Categories",
        sub_cat=sub_cat or "All Sub-Categories",
        plan_type=plan_type or "All Plans",
        option_type=option_type or "All Options",
        start_date=start,
        end_date=end,
    )
    result = leaders_service.build_leaders_dataset(df_all, search_query=search)
    return {
        "rows": df_to_records(result["rows"]),
        "excluded_thin_data": result["excluded_thin_data"],
        "total_funds": result["total_funds"],
        "advancers": result["advancers"],
        "decliners": result["decliners"],
        "market_median_return": sanitize_floats(result["market_median_return"]),
        "top_alpha": sanitize_floats(result["top_alpha"]),
        "leading_category": sanitize_floats(result["leading_category"]),
        "med_vol": sanitize_floats(result["med_vol"]),
        "med_ret": sanitize_floats(result["med_ret"]),
        "quadrant_excluded": result["quadrant_excluded"],
    }

File: frontend/lib/api/leaders.ts
import { apiGet } from "./client";

export interface LeaderRow {
  [key: string]: unknown;
  scheme_code: number;
  scheme_name: string;
  display_name: string;
  fund_house: string;
  category: string;
  broad_category: string;
  plan_type: string;
  option_type: string;
  expense_ratio: number | null;
  latest_nav: number | null;
  dist_from_52w_high_pct: number | null;
  period_return_pct: number;
  annualized_vol_pct: number | null;
  n_trading_days: number;
  win_rate_pct: number | null;
  cat_median_return: number;
  cat_alpha_pct: number;
  quartile: number;
  quartile_rank: string;
  quadrant?: string;
  diagnostic_classification: string;
  return_to_risk: number | null;
}

export interface NamedReturn {
  name: string;
  return_pct: number;
}

export interface LeadersResponse {
  rows: LeaderRow[];
  excluded_thin_data: number;
  total_funds: number;
  advancers: number;
  decliners: number;
  market_median_return: number | null;
  top_alpha: NamedReturn | null;
  leading_category: NamedReturn | null;
  med_vol: number | null;
  med_ret: number | null;
  quadrant_excluded: number;
  vol_lookback?: string;
  vol_methodology?: string;
}

export const getLeaders = (params: {
  broad_cat?: string;
  sub_cat?: string;
  plan_type?: string;
  option_type?: string;
  search?: string;
  start?: string;
  end?: string;
  vol_lookback?: string;
}) => apiGet<LeadersResponse>("/api/leaders", params);
